### PR TITLE
Use variants for underline dotted.

### LINKED ISSUE #4349: Take column into account when rendering cells for perfect patterned underline rendering
Forking off from https://github.com/xtermjs/xterm.js/issues/4060

> Thinking about this and the fact that I want xterm.js rendering to be as great as possible, I think we should support multiple glyphs depending on the x-axis when necessary. For the dotted line specifically, say we had a dot of size `window.devicePixelRatio` with the same sized gap in-between, so there would be at most 2 possible variants for every glyph:
> 
> ```
> Even:
> |* * * * |
> 
> Odd:
> |* * * * *|
> | * * * * |
> ```
> 
> To store this as a unique glyph we need to make the glyph key different, this could be done by asking what underline variant it is and then store it as a number, maybe capped at 4 (2 bits) or 8 variants (3 bits)?
> 
> This would allow us make a non-repeated curvy line and not restrict us to a single cell for both the up and down parts of the wave, if that ends up looking better of course.

Current:

image
image

Desired:

image
image

### PR DESCRIPTION
**Pull request by tisilent**
fix #4349

### CODE REVIEW

**Comment by tisilent** [comment]
hi @Tyriar , The effect should be better this time.

File: src/common/MultiKeyMap.ts [hunk 1]
```diff
@@ -40,3 +40,22 @@ export class FourKeyMap<TFirst extends string | number, TSecond extends string |
     this._data.clear();
   }
 }
+
+export class FiveKeyMap<TFirst extends string | number, TSecond extends string | number, TThird extends string | number, TFourth extends string | number, TFive extends string | number, TValue> {
```
Comment (Tyriar) [inline src/common/MultiKeyMap.ts:44]: Think this can be removed

**Review by Tyriar — CHANGES_REQUESTED** [review]
Thanks! I'll try get to a proper review in a week or 2

**Comment by tisilent** [comment]
I need to clean up the code.
I want to fix the dashed line as well.

**Comment by tisilent** [comment]
Thanks.

File: addons/xterm-addon-canvas/src/TextRenderLayer.ts [hunk 2]
```diff
@@ -72,13 +73,20 @@ export class TextRenderLayer extends BaseRenderLayer {
...
-      y: number
+      y: number,
+      variantOffset: number
     ) => void
   ): void {
+    const fontSize = this._optionsService.rawOptions.fontSize;
+    const drp = this._coreBrowserService.dpr;
+    const lineWidth = Math.max(1, Math.floor(fontSize * drp / 15));
+    const deviceCellWidth = this._charAtlas?.getDeviceCellWidth();
+    let variantOffset: number = -1;
```
Comment (Tyriar) [inline addons/xterm-addon-canvas/src/TextRenderLayer.ts:84]: Can all this logic live inside CellColorResolver?

File: addons/xterm-addon-webgl/src/WebglRenderer.ts [hunk 3]
```diff
@@ -408,11 +410,18 @@ export class WebglRenderer extends Disposable implements IRenderer {
     this._model.cursor = undefined;
     let modelUpdated = false;
 
+    const fontSize = this._optionsService.rawOptions.fontSize;
+    const drp = this._coreBrowserService.dpr;
+    const lineWidth = Math.max(1, Math.floor(fontSize * drp / 15));
+    const deviceCellWidth = this._charAtlas?.getDeviceCellWidth();
```
Comment (Tyriar) [inline addons/xterm-addon-webgl/src/WebglRenderer.ts:416]: Same; can this be shared inside CellColorResolver?

File: addons/xterm-addon-webgl/src/WebglRenderer.ts [hunk 4]
```diff
@@ -408,11 +410,18 @@ export class WebglRenderer extends Disposable implements IRenderer {
...
+
```
Comment (Tyriar) [inline addons/xterm-addon-webgl/src/WebglRenderer.ts:417]: You should be able to pull device cell width off `this.dimensions``

**Comment by tisilent** [comment]
ok.I have discovered an issue now. I will solve it first

File: src/browser/renderer/shared/CellColorResolver.ts [hunk 5]
```diff
@@ -133,5 +147,27 @@ export class CellColorResolver {
...
+        const fontSize = this._terminal.options.fontSize;
+        const drp = this._coreBrowserService.dpr;
+        const lineWidth = Math.max(1, Math.floor(fontSize! * drp / 15));
+        let chWidth: number;
+        if (typeof code === 'number') {
+          chWidth = this._unicodeService.wcwidth(code);
+        } else {
+          chWidth = this._unicodeService.getStringCellWidth(code);
+        }
+        $variantOffset = computeVarinatOffset(deviceCellWidth * chWidth, lineWidth, $variantOffset);
```
Comment (Tyriar) [inline src/browser/renderer/shared/CellColorResolver.ts:167]: I was expecting the variant to be based on the left-most pixel of the glyph, so `chWidth` wouldn't be needed?

File: src/browser/renderer/shared/CellColorResolver.ts [hunk 6]
```diff
@@ -133,5 +147,27 @@ export class CellColorResolver {
...
+    // Reset overrides variantOffset
+    this.result.ext &= ~ExtFlags.VARIANT_OFFSET;
+    this.result.ext |= ($variantOffset << 29) & ExtFlags.VARIANT_OFFSET;
+
+    // Compute next varinatOffset
+    if (cell.extended.underlineStyle === UnderlineStyle.DOTTED) {
+      if (code !== NULL_CELL_CODE) {
+        const fontSize = this._terminal.options.fontSize;
+        const drp = this._coreBrowserService.dpr;
+        const lineWidth = Math.max(1, Math.floor(fontSize! * drp / 15));
```
Comment (Tyriar) [inline src/browser/renderer/shared/CellColorResolver.ts:160]: If you pass in `IOptionsService` you can use `rawOptions.fontSize` and avoid the unsafe `!`

Comment (Tyriar) [inline src/browser/renderer/shared/CellColorResolver.ts:160]:
```suggestion
        const lineWidth = Math.max(1, Math.floor(fontSize! * this._coreBrowserService.dpr / 15));
```

File: src/browser/renderer/shared/CellColorResolver.ts [hunk 7]
```diff
@@ -133,5 +147,27 @@ export class CellColorResolver {
     // Use the override if it exists
     this.result.bg = $hasBg ? $bg : this.result.bg;
     this.result.fg = $hasFg ? $fg : this.result.fg;
+
+    // Reset overrides variantOffset
+    this.result.ext &= ~ExtFlags.VARIANT_OFFSET;
+    this.result.ext |= ($variantOffset << 29) & ExtFlags.VARIANT_OFFSET;
+
+    // Compute next varinatOffset
```
Comment (Tyriar) [inline src/browser/renderer/shared/CellColorResolver.ts:155]:
```suggestion
    // Compute next variantOffset
```

File: src/browser/renderer/shared/TextureAtlas.ts [hunk 8]
```diff
@@ -596,9 +597,28 @@ export class TextureAtlas implements ITextureAtlas {
...
-            this._tmpCtx.setLineDash([Math.round(lineWidth), Math.round(lineWidth)]);
-            this._tmpCtx.moveTo(xChLeft, yTop);
-            this._tmpCtx.lineTo(xChRight, yTop);
+            const offsetWidth = nextOffset === 0 ? 0 :
+              (nextOffset >= lineWidth ? lineWidth * 2 - nextOffset : lineWidth - nextOffset);
+            if (offsetWidth === 0) {
+              this._tmpCtx.setLineDash([Math.round(lineWidth), Math.round(lineWidth)]);
+              this._tmpCtx.moveTo(xChLeft, yTop);
+              this._tmpCtx.lineTo(xChRight, yTop);
+            } else {
```
Comment (Tyriar) [inline src/browser/renderer/shared/TextureAtlas.ts:606]: I don't think you need a special case for this as the main effect is `+ offsetWidth` = `+ 0`?

File: src/common/buffer/AttributeData.ts [hunk 9]
```diff
@@ -174,6 +177,18 @@ export class ExtendedAttrs implements IExtendedAttrs {
     this._urlId = value;
   }
 
+  public get underlineVarinatOffset(): number {
+    const val = (this._ext & ExtFlags.VARIANT_OFFSET) >> 29;
+    if (val < 0) {
+      return val ^ 4294967288;
```
Comment (Tyriar) [inline src/common/buffer/AttributeData.ts:183]: Hex would be easier to read here

**Review by Tyriar — CHANGES_REQUESTED** [review]

File: src/browser/renderer/shared/CellColorResolver.ts (hunk 5 above)
Comment (tisilent) [reply src/browser/renderer/shared/CellColorResolver.ts:167]: I need to take a look at the process here.....

Comment (tisilent) [reply src/browser/renderer/shared/CellColorResolver.ts:166]: I see. 😄

**Review by Tyriar — APPROVED** [review]
This works really well! 👏 

image

image

image

**Comment by tisilent** [comment]
Next. Curve.

**Comment by Tyriar** [comment]
@tisilent that would be great, I found it a bit tricky to get the curve to look good at various zoom levels.

**Pull request merged**